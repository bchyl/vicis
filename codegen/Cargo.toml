[package]
name = "vicis-codegen"
version = "0.2.0"
authors = ["uint256_t"]
edition = "2021"
license-file = "../LICENSE"
description = "Manipulate LLVM-IR in Pure Rust"
homepage = "https://github.com/maekawatoshiki/vicis"
documentation = "https://docs.rs/vicis"
repository = "https://github.com/maekawatoshiki/vicis"
readme = "../README.md"
keywords = ["llvm", "compiler"]

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
vicis-core = { path = "../core" }
rustc-hash = "^1.1.0"
id-arena = "^2.2.1"
anyhow = "^1.0.38"
log = "0.4.17"

[dev-dependencies]
color-backtrace = "0.5.1"
env_logger = "0.9.0"
insta = "^1.7.1"
structopt = "0.3.26"
