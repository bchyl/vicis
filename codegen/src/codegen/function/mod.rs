pub mod basic_block;
pub mod data;
pub mod instruction;
pub mod layout;
pub mod slot;

use super::{call_conv::CallConvKind, isa::TargetIsa};
use crate::codegen::function::instruction::InstructionInfo;
use instruction::InstructionId;
use std::{fmt, marker::PhantomData};
use vicis_core::ir::{function::Function as IrFunction, types::Types};

pub struct Function<'a, T: TargetIsa> {
    pub ir: &'a IrFunction,
    pub data: data::Data<<T::InstInfo as InstructionInfo>::Data>,
    pub layout: layout::Layout<<T::InstInfo as InstructionInfo>::Data>,
    pub slots: slot::Slots<T>,
    pub types: Types,
    pub is_declaration: bool,
    pub call_conv: CallConvKind,
    pub _isa: PhantomData<fn() -> T>,
}

impl<T: TargetIsa> Function<'_, T> {
    pub fn remove_inst(
        &mut self,
        inst: InstructionId<<T::InstInfo as InstructionInfo>::Data>,
    ) -> Option<()> {
        self.layout.remove_inst(inst)
    }
}

impl<T: TargetIsa> fmt::Debug for Function<'_, T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        if self.is_declaration {
            write!(f, "declare ")?
        } else {
            write!(f, "define ")?
        }
        write!(f, "{:?} ", self.ir.preemption_specifier)?;
        write!(f, "{} ", self.types.to_string(self.ir.result_ty))?;
        write!(f, "@{}(", self.ir.name)?;
        for (i, param) in self.ir.params.iter().enumerate() {
            write!(
                f,
                "{} %A{}{}",
                self.types.to_string(param.ty),
                i,
                if i == self.ir.params.len() - 1 {
                    ""
                } else {
                    ", "
                }
            )?;
        }
        write!(f, ") ")?;
        for attr in &self.ir.func_attrs {
            write!(f, "{:?}", attr)?;
        }

        if self.is_declaration {
            writeln!(f)?;
        } else {
            writeln!(f, "{{")?;
            for block_id in self.layout.block_iter() {
                writeln!(f, "B{:?}:", block_id.index())?;
                for inst_id in self.layout.inst_iter(block_id) {
                    let inst = self.data.inst_ref(inst_id);
                    writeln!(f, "  id{:<4}| {:?}", inst_id.index(), inst.data)?;
                }
            }
            writeln!(f, "}}")?;
        }

        Ok(())
    }
}
