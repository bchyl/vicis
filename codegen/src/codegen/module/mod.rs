use super::{function::Function, isa::TargetIsa};
use id_arena::Arena;
use rustc_hash::FxHashMap;
use std::{fmt, marker::PhantomData};
use vicis_core::ir::{
    module::{
        attributes::Attribute, global_variable::GlobalVariable, name::Name, Module as IrModule,
    },
    types::Types,
};

pub struct Module<'a, T: TargetIsa> {
    pub module: &'a IrModule,
    pub functions: Arena<Function<'a, T>>,
    pub attributes: FxHashMap<u32, Vec<Attribute>>,
    pub global_variables: FxHashMap<Name, GlobalVariable>,
    pub types: Types,
    pub _isa: PhantomData<fn() -> T>,
}

impl<T: TargetIsa> fmt::Debug for Module<'_, T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "source_filename = \"{}\"", self.module.source_filename())?;
        writeln!(
            f,
            "target datalayout = \"{}\"",
            self.module.target().datalayout()
        )?;
        writeln!(f, "target triple = \"{}\"", self.module.target().triple())?;
        writeln!(f)?;
        for gv in self.global_variables.values() {
            writeln!(f, "{}", gv.to_string(&self.types))?;
        }
        writeln!(f)?;
        for (_, func) in &self.functions {
            writeln!(f, "{:?}", func)?;
        }
        for (id, attrs) in &self.attributes {
            write!(f, "attributes #{} = {{ ", id)?;
            for attr in attrs {
                write!(f, "{:?} ", attr)?;
            }
            writeln!(f, "}}")?
        }
        Ok(())
    }
}
