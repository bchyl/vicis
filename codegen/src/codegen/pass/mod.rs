pub mod liveness;
pub mod regalloc;
pub mod spiller;
