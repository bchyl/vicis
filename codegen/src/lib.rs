#![feature(let_chains)]

extern crate vicis_core;

#[macro_use]
pub mod macros;
pub mod codegen;
