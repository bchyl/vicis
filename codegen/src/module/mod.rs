use super::{function::Function, isa::TargetIsa};
use id_arena::Arena;
use std::fmt;
use vicis_core::ir::{module::Module as IrModule, types::Types};

pub struct Module<'a, T: TargetIsa> {
    pub ir: &'a IrModule,
    pub functions: Arena<Function<'a, T>>,
    pub types: Types,
    pub isa: &'a T,
}

impl<T: TargetIsa> fmt::Debug for Module<'_, T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "source_filename = \"{}\"", self.ir.source_filename())?;
        writeln!(
            f,
            "target datalayout = \"{}\"",
            self.ir.target().datalayout()
        )?;
        writeln!(f, "target triple = \"{}\"", self.ir.target().triple())?;
        writeln!(f)?;
        for gv in self.ir.global_variables().values() {
            writeln!(f, "{}", gv.to_string(&self.types))?;
        }
        writeln!(f)?;
        for (_, func) in &self.functions {
            writeln!(f, "{:?}", func)?;
        }
        for (id, attrs) in self.ir.attributes() {
            write!(f, "attributes #{} = {{ ", id)?;
            for attr in attrs {
                write!(f, "{:?} ", attr)?;
            }
            writeln!(f, "}}")?
        }
        Ok(())
    }
}
