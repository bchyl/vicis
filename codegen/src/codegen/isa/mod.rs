pub mod x86_64;

use crate::codegen::{
    call_conv::CallConvKind,
    function::instruction::TargetInst,
    lower,
    module::Module,
    register::{RegisterClass, RegisterInfo},
};
use anyhow::Result;
use vicis_core::ir::{
    module::data_layout::DataLayout,
    types::{Type, Types},
};

pub trait TargetIsa: Clone {
    type Inst: TargetInst;
    type RegClass: RegisterClass;
    type RegInfo: RegisterInfo;
    type Lower: lower::Lower<Self>;

    fn module_passes() -> Vec<for<'a, 'b> fn(&'b mut Module<'a, Self>) -> Result<()>>; // TODO: Implement a pass manager for machine modules
    fn default_call_conv() -> CallConvKind;
    fn type_size(types: &Types, ty: Type) -> u32; // TODO: FIXME: DataLayout can replace this.
    fn data_layout(&self) -> &DataLayout;
}
