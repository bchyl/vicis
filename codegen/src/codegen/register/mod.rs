use crate::codegen::{
    call_conv::CallConvKind,
    function::instruction::{InstructionData as ID, InstructionId},
};
use rustc_hash::FxHashMap;
use vicis_core::ir::types::{Type, Types};

#[derive(Debug, Clone, Copy, Eq, Hash, PartialEq)]
pub struct Reg(pub u16, pub u16);

#[derive(Debug, Clone, Copy, Eq, Hash, PartialEq)]
pub struct RegUnit(pub u16, pub u16); // Same as top-level register

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct VReg(pub u32);

#[derive(Default)]
pub struct VRegs {
    pub map: FxHashMap<VReg, VRegData>,
    pub cur: u32,
}

pub struct VRegData {
    pub vreg: VReg,
    pub ty: Type,
    // ...
}

pub struct VRegUsers<Data: ID> {
    pub vreg_to_insts: FxHashMap<VReg, Vec<VRegUser<Data>>>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct VRegUser<Data: ID> {
    pub inst_id: InstructionId<Data>,
    pub read: bool,
    pub write: bool,
}

pub trait RegisterInfo {
    fn arg_reg_list(cc: &CallConvKind) -> &'static [RegUnit];
    fn to_reg_unit(reg: Reg) -> RegUnit;
}

pub trait RegisterClass {
    fn for_type(types: &Types, id: Type) -> Self;
    fn gpr_list(&self) -> Vec<Reg>;
    fn apply_for(&self, ru: RegUnit) -> Reg;
}

impl RegUnit {
    pub fn apply<RC: RegisterClass>(self, rc: &RC) -> Reg {
        rc.apply_for(self)
    }
}

impl VRegs {
    pub fn new() -> Self {
        Self::default()
    }

    // TODO: Change name
    pub fn add_vreg_data(&mut self, ty: Type) -> VReg {
        let key = VReg(self.cur);
        self.map.insert(key, VRegData { vreg: key, ty });
        self.cur += 1;
        key
    }

    pub fn create_from(&mut self, vreg: VReg) -> VReg {
        let ty = self.map[&vreg].ty;
        self.add_vreg_data(ty)
    }

    pub fn type_for(&self, vreg: VReg) -> Type {
        self.map[&vreg].ty
    }

    pub fn change_ty(&mut self, vreg: VReg, ty: Type) {
        self.map.get_mut(&vreg).unwrap().ty = ty
    }
}

impl<Data: ID> Default for VRegUsers<Data> {
    fn default() -> Self {
        Self {
            vreg_to_insts: FxHashMap::default(),
        }
    }
}

impl<Data: ID> VRegUsers<Data> {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn add_use(&mut self, vreg: VReg, inst_id: InstructionId<Data>, read: bool, write: bool) {
        self.vreg_to_insts
            .entry(vreg)
            .or_insert_with(Vec::new)
            .push(VRegUser {
                inst_id,
                read,
                write,
            })
    }

    pub fn get(&self, vreg: VReg) -> &Vec<VRegUser<Data>> {
        &self.vreg_to_insts[&vreg]
    }

    pub fn remove_use(
        &mut self,
        vreg: VReg,
        inst_id: InstructionId<Data>,
    ) -> Option<VRegUser<Data>> {
        if let Some(idx) = self.vreg_to_insts[&vreg]
            .iter()
            .position(|u| u.inst_id == inst_id)
        {
            return Some(self.vreg_to_insts.get_mut(&vreg).unwrap().remove(idx));
        }
        None
    }
}
