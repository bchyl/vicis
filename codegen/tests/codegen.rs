use std::fs;
use vicis_codegen::codegen::{
    isa::x86_64::X86_64, lower::compile_module, module::Module as MachModule,
};
use vicis_core::ir::module::Module;

macro_rules! test {
    ($testname:ident, $name:expr) => {
        #[test]
        fn $testname() {
            let mach_module = compile($name);
            insta::assert_display_snapshot!(mach_module);
        }
    };
}

test!(test_add, "ary1");
test!(test_ary2, "ary2");
test!(test_ary3, "ary3");
test!(test_ary4, "ary4");
test!(test_ary5, "ary5");
test!(test_br, "br");
test!(test_call1, "call1");
test!(test_call2, "call2");
test!(test_condbr, "condbr");
test!(test_fibo, "fibo");
test!(test_load_add, "load_add");
test!(test_phi, "phi");
test!(test_phi2, "phi2");
test!(test_puts, "puts");
test!(test_sum, "sum");

#[cfg(test)]
fn compile(name: &str) -> MachModule<X86_64> {
    let parent = "./tests/codegen/";
    let input = format!("{}{}.ll", parent, name);
    let input_body = &fs::read_to_string(input).unwrap();
    let module = Module::try_from(input_body.as_str()).unwrap();
    compile_module(X86_64, &module).unwrap()
}
