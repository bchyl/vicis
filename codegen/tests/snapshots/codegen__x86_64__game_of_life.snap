---
source: codegen/tests/codegen.rs
assertion_line: 45
expression: mach_module.display_asm()
---
  .text
  .intel_syntax noprefix
  .text
  .globl main
main:
.LBL0_0:
  push rbp
  mov rbp, rsp
  sub rsp, 3360
  mov dword ptr [rbp-4], 0
  lea rdi, [rbp-3280]
  lea rsi, [rip + __const.main.grid]
  mov rdx, 1600
  call memcpy
  mov dword ptr [rbp-3292], 0
  jmp .LBL0_1
.LBL0_1:
  mov eax, dword ptr [rbp-3292]
  cmp eax, 50
  jl .LBL0_2
  jmp .LBL0_31
.LBL0_2:
  lea rdi, [rip + .str]
  call printf
  mov dword ptr [rbp-3284], 0
  jmp .LBL0_3
.LBL0_3:
  mov eax, dword ptr [rbp-3284]
  cmp eax, 20
  jl .LBL0_4
  jmp .LBL0_13
.LBL0_4:
  lea rdi, [rip + .str.1]
  call printf
  mov dword ptr [rbp-3288], 0
  jmp .LBL0_5
.LBL0_5:
  mov eax, dword ptr [rbp-3288]
  cmp eax, 20
  jl .LBL0_6
  jmp .LBL0_11
.LBL0_6:
  movsxd rax, dword ptr [rbp-3284]
  movsxd rcx, dword ptr [rbp-3288]
  lea rdx, [rbp-3280]
  imul rax, rax, 80
  add rdx, rax
  mov eax, dword ptr [rdx+0+rcx*4]
  cmp eax, 1
  je .LBL0_7
  jmp .LBL0_8
.LBL0_7:
  lea rdi, [rip + .str.2]
  call printf
  jmp .LBL0_9
.LBL0_8:
  lea rdi, [rip + .str.3]
  call printf
  jmp .LBL0_9
.LBL0_9:
  mov esi, dword ptr [rbp-3284]
  mov edx, dword ptr [rbp-3288]
  lea rdi, [rbp-3280]
  mov ecx, 20
  call count_nbr
  movsxd rcx, dword ptr [rbp-3284]
  movsxd rdx, dword ptr [rbp-3288]
  lea rdi, [rbp-1616]
  imul rcx, rcx, 80
  add rdi, rcx
  mov dword ptr [rdi+0+rdx*4], eax
  jmp .LBL0_10
.LBL0_10:
  mov eax, dword ptr [rbp-3288]
  add eax, 1
  mov dword ptr [rbp-3288], eax
  jmp .LBL0_5
.LBL0_11:
  jmp .LBL0_12
.LBL0_12:
  mov eax, dword ptr [rbp-3284]
  add eax, 1
  mov dword ptr [rbp-3284], eax
  jmp .LBL0_3
.LBL0_13:
  mov dword ptr [rbp-3284], 0
  jmp .LBL0_14
.LBL0_14:
  mov eax, dword ptr [rbp-3284]
  cmp eax, 20
  jl .LBL0_15
  jmp .LBL0_29
.LBL0_15:
  mov dword ptr [rbp-3288], 0
  jmp .LBL0_16
.LBL0_16:
  mov eax, dword ptr [rbp-3288]
  cmp eax, 20
  jl .LBL0_17
  jmp .LBL0_27
.LBL0_17:
  movsxd rax, dword ptr [rbp-3284]
  movsxd rcx, dword ptr [rbp-3288]
  lea rdx, [rbp-3280]
  imul rax, rax, 80
  add rdx, rax
  mov eax, dword ptr [rdx+0+rcx*4]
  cmp eax, 1
  jge .LBL0_18
  jmp .LBL0_22
.LBL0_18:
  movsxd rax, dword ptr [rbp-3284]
  movsxd rcx, dword ptr [rbp-3288]
  lea rdx, [rbp-1616]
  imul rax, rax, 80
  add rdx, rax
  mov eax, dword ptr [rdx+0+rcx*4]
  cmp eax, 1
  jle .LBL0_20
  jmp .LBL0_19
.LBL0_19:
  movsxd rax, dword ptr [rbp-3284]
  movsxd rcx, dword ptr [rbp-3288]
  lea rdx, [rbp-1616]
  imul rax, rax, 80
  add rdx, rax
  mov eax, dword ptr [rdx+0+rcx*4]
  cmp eax, 4
  jge .LBL0_20
  jmp .LBL0_21
.LBL0_20:
  movsxd rax, dword ptr [rbp-3284]
  movsxd rcx, dword ptr [rbp-3288]
  lea rdx, [rbp-3280]
  imul rax, rax, 80
  add rdx, rax
  mov dword ptr [rdx+0+rcx*4], 0
  jmp .LBL0_21
.LBL0_21:
  jmp .LBL0_25
.LBL0_22:
  movsxd rax, dword ptr [rbp-3284]
  movsxd rcx, dword ptr [rbp-3288]
  lea rdx, [rbp-1616]
  imul rax, rax, 80
  add rdx, rax
  mov eax, dword ptr [rdx+0+rcx*4]
  cmp eax, 3
  je .LBL0_23
  jmp .LBL0_24
.LBL0_23:
  movsxd rax, dword ptr [rbp-3284]
  movsxd rcx, dword ptr [rbp-3288]
  lea rdx, [rbp-3280]
  imul rax, rax, 80
  add rdx, rax
  mov dword ptr [rdx+0+rcx*4], 1
  jmp .LBL0_24
.LBL0_24:
  jmp .LBL0_25
.LBL0_25:
  jmp .LBL0_26
.LBL0_26:
  mov eax, dword ptr [rbp-3288]
  add eax, 1
  mov dword ptr [rbp-3288], eax
  jmp .LBL0_16
.LBL0_27:
  jmp .LBL0_28
.LBL0_28:
  mov eax, dword ptr [rbp-3284]
  add eax, 1
  mov dword ptr [rbp-3284], eax
  jmp .LBL0_14
.LBL0_29:
  mov edi, 100000
  call usleep
  jmp .LBL0_30
.LBL0_30:
  mov eax, dword ptr [rbp-3292]
  add eax, 1
  mov dword ptr [rbp-3292], eax
  jmp .LBL0_1
.LBL0_31:
  mov eax, 0
  add rsp, 3360
  pop rbp
  ret 
  .text
  .globl count_nbr
count_nbr:
.LBL3_0:
  push rbp
  mov rbp, rsp
  sub rsp, 32
  mov rax, rdi
  mov edi, esi
  mov qword ptr [rbp-8], rax
  mov dword ptr [rbp-12], edi
  mov dword ptr [rbp-16], edx
  mov dword ptr [rbp-20], ecx
  mov dword ptr [rbp-24], 0
  mov eax, dword ptr [rbp-12]
  sub eax, 1
  cmp eax, 0
  jge .LBL3_1
  jmp .LBL3_5
.LBL3_1:
  mov eax, dword ptr [rbp-16]
  sub eax, 1
  cmp eax, 0
  jge .LBL3_2
  jmp .LBL3_5
.LBL3_2:
  mov rax, qword ptr [rbp-8]
  mov ecx, dword ptr [rbp-12]
  mov edx, dword ptr [rbp-16]
  sub ecx, 1
  movsxd rcx, ecx
  lea rax, [rax]
  imul rcx, rcx, 80
  add rax, rcx
  mov ecx, edx
  sub ecx, 1
  movsxd rcx, ecx
  mov eax, dword ptr [rax+0+rcx*4]
  cmp eax, 1
  jge .LBL3_3
  jmp .LBL3_4
.LBL3_3:
  mov eax, dword ptr [rbp-24]
  add eax, 1
  mov dword ptr [rbp-24], eax
  jmp .LBL3_4
.LBL3_4:
  jmp .LBL3_5
.LBL3_5:
  mov eax, dword ptr [rbp-12]
  sub eax, 1
  cmp eax, 0
  jge .LBL3_6
  jmp .LBL3_9
.LBL3_6:
  mov rax, qword ptr [rbp-8]
  mov ecx, dword ptr [rbp-12]
  movsxd rdx, dword ptr [rbp-16]
  sub ecx, 1
  movsxd rcx, ecx
  lea rax, [rax]
  imul rcx, rcx, 80
  add rax, rcx
  mov eax, dword ptr [rax+0+rdx*4]
  cmp eax, 1
  jge .LBL3_7
  jmp .LBL3_8
.LBL3_7:
  mov eax, dword ptr [rbp-24]
  add eax, 1
  mov dword ptr [rbp-24], eax
  jmp .LBL3_8
.LBL3_8:
  jmp .LBL3_9
.LBL3_9:
  mov eax, dword ptr [rbp-12]
  sub eax, 1
  cmp eax, 0
  jge .LBL3_10
  jmp .LBL3_14
.LBL3_10:
  mov eax, dword ptr [rbp-16]
  mov ecx, dword ptr [rbp-20]
  add eax, 1
  cmp eax, ecx
  jl .LBL3_11
  jmp .LBL3_14
.LBL3_11:
  mov rax, qword ptr [rbp-8]
  mov ecx, dword ptr [rbp-12]
  mov edx, dword ptr [rbp-16]
  sub ecx, 1
  movsxd rcx, ecx
  lea rax, [rax]
  imul rcx, rcx, 80
  add rax, rcx
  mov ecx, edx
  add ecx, 1
  movsxd rcx, ecx
  mov eax, dword ptr [rax+0+rcx*4]
  cmp eax, 1
  jge .LBL3_12
  jmp .LBL3_13
.LBL3_12:
  mov eax, dword ptr [rbp-24]
  add eax, 1
  mov dword ptr [rbp-24], eax
  jmp .LBL3_13
.LBL3_13:
  jmp .LBL3_14
.LBL3_14:
  mov eax, dword ptr [rbp-16]
  sub eax, 1
  cmp eax, 0
  jge .LBL3_15
  jmp .LBL3_18
.LBL3_15:
  mov rax, qword ptr [rbp-8]
  movsxd rcx, dword ptr [rbp-12]
  mov edx, dword ptr [rbp-16]
  lea rax, [rax]
  imul rcx, rcx, 80
  add rax, rcx
  mov ecx, edx
  sub ecx, 1
  movsxd rcx, ecx
  mov eax, dword ptr [rax+0+rcx*4]
  cmp eax, 1
  jge .LBL3_16
  jmp .LBL3_17
.LBL3_16:
  mov eax, dword ptr [rbp-24]
  add eax, 1
  mov dword ptr [rbp-24], eax
  jmp .LBL3_17
.LBL3_17:
  jmp .LBL3_18
.LBL3_18:
  mov eax, dword ptr [rbp-16]
  mov ecx, dword ptr [rbp-20]
  add eax, 1
  cmp eax, ecx
  jl .LBL3_19
  jmp .LBL3_22
.LBL3_19:
  mov rax, qword ptr [rbp-8]
  movsxd rcx, dword ptr [rbp-12]
  mov edx, dword ptr [rbp-16]
  lea rax, [rax]
  imul rcx, rcx, 80
  add rax, rcx
  mov ecx, edx
  add ecx, 1
  movsxd rcx, ecx
  mov eax, dword ptr [rax+0+rcx*4]
  cmp eax, 1
  jge .LBL3_20
  jmp .LBL3_21
.LBL3_20:
  mov eax, dword ptr [rbp-24]
  add eax, 1
  mov dword ptr [rbp-24], eax
  jmp .LBL3_21
.LBL3_21:
  jmp .LBL3_22
.LBL3_22:
  mov eax, dword ptr [rbp-12]
  mov ecx, dword ptr [rbp-20]
  add eax, 1
  cmp eax, ecx
  jl .LBL3_23
  jmp .LBL3_27
.LBL3_23:
  mov eax, dword ptr [rbp-16]
  sub eax, 1
  cmp eax, 0
  jge .LBL3_24
  jmp .LBL3_27
.LBL3_24:
  mov rax, qword ptr [rbp-8]
  mov ecx, dword ptr [rbp-12]
  mov edx, dword ptr [rbp-16]
  add ecx, 1
  movsxd rcx, ecx
  lea rax, [rax]
  imul rcx, rcx, 80
  add rax, rcx
  mov ecx, edx
  sub ecx, 1
  movsxd rcx, ecx
  mov eax, dword ptr [rax+0+rcx*4]
  cmp eax, 1
  jge .LBL3_25
  jmp .LBL3_26
.LBL3_25:
  mov eax, dword ptr [rbp-24]
  add eax, 1
  mov dword ptr [rbp-24], eax
  jmp .LBL3_26
.LBL3_26:
  jmp .LBL3_27
.LBL3_27:
  mov eax, dword ptr [rbp-12]
  mov ecx, dword ptr [rbp-20]
  add eax, 1
  cmp eax, ecx
  jl .LBL3_28
  jmp .LBL3_31
.LBL3_28:
  mov rax, qword ptr [rbp-8]
  mov ecx, dword ptr [rbp-12]
  movsxd rdx, dword ptr [rbp-16]
  add ecx, 1
  movsxd rcx, ecx
  lea rax, [rax]
  imul rcx, rcx, 80
  add rax, rcx
  mov eax, dword ptr [rax+0+rdx*4]
  cmp eax, 1
  jge .LBL3_29
  jmp .LBL3_30
.LBL3_29:
  mov eax, dword ptr [rbp-24]
  add eax, 1
  mov dword ptr [rbp-24], eax
  jmp .LBL3_30
.LBL3_30:
  jmp .LBL3_31
.LBL3_31:
  mov eax, dword ptr [rbp-12]
  mov ecx, dword ptr [rbp-20]
  add eax, 1
  cmp eax, ecx
  jl .LBL3_32
  jmp .LBL3_36
.LBL3_32:
  mov eax, dword ptr [rbp-16]
  mov ecx, dword ptr [rbp-20]
  add eax, 1
  cmp eax, ecx
  jl .LBL3_33
  jmp .LBL3_36
.LBL3_33:
  mov rax, qword ptr [rbp-8]
  mov ecx, dword ptr [rbp-12]
  mov edx, dword ptr [rbp-16]
  add ecx, 1
  movsxd rcx, ecx
  lea rax, [rax]
  imul rcx, rcx, 80
  add rax, rcx
  mov ecx, edx
  add ecx, 1
  movsxd rcx, ecx
  mov eax, dword ptr [rax+0+rcx*4]
  cmp eax, 1
  jge .LBL3_34
  jmp .LBL3_35
.LBL3_34:
  mov eax, dword ptr [rbp-24]
  add eax, 1
  mov dword ptr [rbp-24], eax
  jmp .LBL3_35
.LBL3_35:
  jmp .LBL3_36
.LBL3_36:
  mov eax, dword ptr [rbp-24]
  add rsp, 32
  pop rbp
  ret 
__const.main.grid:
  .zero 80
  .zero 80
  .zero 80
  .zero 80
  .zero 80
  .zero 80
  .zero 80
  .long 0
  .long 0
  .long 0
  .long 0
  .long 0
  .long 1
  .long 1
  .long 1
  .long 1
  .long 1
  .long 1
  .long 1
  .long 1
  .long 1
  .long 1
  .long 0
  .long 0
  .long 0
  .long 0
  .long 0
  .zero 80
  .zero 80
  .zero 80
  .zero 80
  .zero 80
  .zero 80
  .long 0
  .long 0
  .long 1
  .long 1
  .long 1
  .long 1
  .zero 56
  .long 0
  .long 1
  .long 0
  .long 0
  .long 0
  .long 1
  .zero 56
  .long 0
  .long 0
  .long 0
  .long 0
  .long 0
  .long 1
  .zero 56
  .long 0
  .long 1
  .long 0
  .long 0
  .long 1
  .zero 60
  .zero 80
  .zero 80
.str.1:
  .string "\n"
.str:
  .string "\x1b[0;0H"
.str.2:
  .string "\x1b[42m  \x1b[m"
.str.3:
  .string "\x1b[47m  \x1b[m"

